crates/trigger-http2/src/outbound_http.rs

```diff
@@ -0,0 +1,64 @@
+use std::{
+    net::{IpAddr, Ipv4Addr, SocketAddr},
+    sync::Arc,
+};
+
+use spin_factor_outbound_http::{HostFutureIncomingResponse, SelfRequestOrigin};
+use spin_http::routes::RouteMatch;
+use spin_outbound_networking::parse_service_chaining_target;
+use wasmtime_wasi_http::types::IncomingResponse;
+
+use crate::server::HttpServer;
+
+pub struct OutboundHttpInterceptor {
+    server: Arc<HttpServer>,
+    origin: SelfRequestOrigin,
+}
+
+impl OutboundHttpInterceptor {
+    pub fn new(server: Arc<HttpServer>, origin: SelfRequestOrigin) -> Self {
+        Self { server, origin }
+    }
+}
+
+const CHAINED_CLIENT_ADDR: SocketAddr = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 0);
+
+impl spin_factor_outbound_http::OutboundHttpInterceptor for OutboundHttpInterceptor {
```

Comment:
I'm surprised to see this remain in the trigger. I had been thinking that factors would allow us to move this logic (which doesn't seem to be trigger specific) closer to the outbound-http handler code.

Reply:
Looking more deeply into this, I think I understand why at least *some* of this has to be in the trigger. The interceptor is meant as a way for the trigger to inject custom logic on each request. In this case, essentially the HTTP trigger's interceptor is responsible for the following things:
1. Establishing the `SelfRequestOrigin` so that self requests can be routed properly.
  1a. The factor has an implicit assumption that somewhere someone is adding a `SelfRequestOrigin` to the request extensions. If no one does, requests to `self` will fail. 
2. Figuring out whether the request is a a service chained request. If it is a service chaining request, then routing that back through the HTTP trigger. 

Both seem to be very heavily tied to the specific HTTP trigger though I'm realizing now, I'm not sure we ever really rationalized these features outside of the HTTP trigger. What does a self request or a service chained request mean when the app is invoked outside of the HTTP trigger? 

As the code is currently written, self requests seem *indirectly* to not be a feature outside of the HTTP trigger. If the factor is invoked outside of an HTTP trigger and no origin is set on the request extensions, then the outgoing request will fail. If a trigger wants to opt into supporting self requests, they have to do so in a very indirect way (by setting the `SelfRequestOrigin` on the request extensions) - this is certainly less than ideal... The outbound-http factor code is very much aware of this feature, but exposes it to embedders in a very indirect way. 

Service chaining, on the other hand, seems to be explicitly an HTTP trigger only feature. If "spin.internal" is used in an outbound request in any other trigger, that request will be routed to "spin.internal" as if it were any other DNS based authority (I guess let's hope that the user is running Spin on a network where `spin.internal` doesn't route to anything malicious...). The outbound-http factor knows nothing about this feature, and it is implemented purely within the HTTP trigger.

Reply:
I explored a couple of different approaches here but decided to follow the existing semantics for pretty much the reasons you mention:

- "origin" is inherently somewhat trigger-specific; we _could_ have a universal e.g. runtime config option to set/override it for any trigger type, but that seemed out of scope for a refactor

- service chaining could be implemented for other trigger types but would require a large chunk of the http trigger be included in the outbound http factor which again seems out of scope

Reply:
All of that sounds good to me - I have another comment about how I don't love how we're feeding the origin information back into the factor, but that's an implementation detail we can change after we merge this PR if we want.